dotnet/src/Connectors/Connectors.Memory.Pinecone/PineconeVectorStoreRecordMapper.cs

```diff
@@ -0,0 +1,196 @@
+            if (propertyValue != null)
+            {
+                metadata[propertyName] = ConvertToMetadataValue(propertyValue);
+            }
+        }
+
+        var vectorPropertyInfo = this._vectorPropertiesInfo.First();
+        var valuesObject = vectorPropertyInfo.GetValue(dataModel);
+        if (valuesObject is not ReadOnlyMemory<float> values)
+        {
+            throw new VectorStoreRecordMappingException($"Vector property {vectorPropertyInfo.Name} on provided record of type {typeof(TRecord).FullName} may not be null.");
+        }
+
+        // TODO: what about sparse values?
+        var result = new Vector
+        {
+            Id = (string)keyObject,
+            Values = values.ToArray(),
+            Metadata = metadata,
+            SparseValues = null
+        };
+
+        return result;
+    }
+
+    /// <inheritdoc />
+    public TRecord MapFromStorageToDataModel(Vector storageModel, StorageToDataModelMapperOptions options)
+    {
+        var keyPropertyName = this._storagePropertyNames[this._keyPropertyInfo.Name];
+
+        var outputJsonObject = new JsonObject
+        {
+            { this._keyPropertyInfo.Name, JsonValue.Create(storageModel.Id) },
+        };
+
+        if (options?.IncludeVectors is true)
+        {
+            var vectorProperty = this._vectorPropertiesInfo.First();
+            var propertyName = this._storagePropertyNames[vectorProperty.Name];
+            outputJsonObject.Add(vectorProperty.Name, new JsonArray(storageModel.Values.Select(x => JsonValue.Create(x)).ToArray()));
+        }
+
+        if (storageModel.Metadata != null)
+        {
+            foreach (var payloadProperty in this._payloadPropertiesInfo)
+            {
+                var propertyName = this._storagePropertyNames[payloadProperty.Name];
+                if (storageModel.Metadata.TryGetValue(propertyName, out var value))
+                {
+                    outputJsonObject.Add(payloadProperty.Name, ConvertFromMetadataValueToJsonNode(value));
+                }
+            }
+        }
+
+        return outputJsonObject.Deserialize<TRecord>()!;
+    }
+
+    private static JsonNode? ConvertFromMetadataValueToJsonNode(MetadataValue metadataValue)
+        => metadataValue.Inner switch
+        {
+            null => null,
+            bool boolValue => JsonValue.Create(boolValue),
+            string stringValue => JsonValue.Create(stringValue),
+            int intValue => JsonValue.Create(intValue),
+            long longValue => JsonValue.Create(longValue),
+            float floatValue => JsonValue.Create(floatValue),
+            double doubleValue => JsonValue.Create(doubleValue),
+            decimal decimalValue => JsonValue.Create(decimalValue),
+            MetadataValue[] array => new JsonArray(array.Select(ConvertFromMetadataValueToJsonNode).ToArray()),
+            _ => throw new VectorStoreRecordMappingException($"Unsupported metadata value: '{metadataValue.Inner}'."),
```

Comment:
Would it be better to specify the type that isn't supported, i.e. `Unsupported metadata type: '{metadataValue.Inner?.GetType().FullName}'`, since the value itself may be sensitive and the type might be more useful to troubleshoot how the fix the problem.